### PR TITLE
Primitive Technology - Paper Filters

### PR DESCRIPTION
**Pull request by fludd12**
image
The very first of (hopefully, but I can't promise anything) several PRs to allow for more functions available to those who want to jury rig equipment from basic parts, especially those on Lavaland.
## About The Pull Request
This PR adds a simple method for filtering certain chemicals out of liquids. You can now craft a filter bowl with two wooden planks. Inserting paper into the bowl creates a blank filter - pouring a reagent container into the bowl with a blank filter installed keys the filter to whatever chemical is most present in the reagent container at the time. From that point on, while the filter is in the bowl, the bowl will only accept that reagent, leaving the others inside the original container.

Filters can be removed with alt click, allowing you to reuse the bowl for a variety of purposes. Additionally, while the bowl lacks a filter, it simply acts as a regular 50u reagent container.

## Why It's Good For The Game
There has been a couple of suggested options to allow for "ghetto" reagent filtering in the past, but none of them have been particularly well received, either due to a lack of simplicity or just a lack of interest. I am hoping the root cause is the former, and that this provides an elegant solution - after all, I assume most people have heard of a coffee filter.

## Changelog
:cl: Fel
add: You can craft a filter bowl with wood, letting you filter certain reagents out of other containers!
/:cl:

### CODE REVIEW

**Comment by SkeletalElite** [comment]
Are there any chems (that someone might reasonably ghetto craft) where it will be difficult to actually key a filter to it because the chem ends being a small amount compared to the rest of the reagents in the container unless you use very precise amounts?

File: code/modules/reagents/reagent_containers/glass.dm [hunk 1]
```diff
@@ -347,6 +347,111 @@
...
+	if(chemfilter)
+		return
+	if(!istype(I))
+		return
+	chemfilter = I
+	reagents.whitelist = list(chemfilter.filterchem)
+	update_icon()
+	I.forceMove(src)
+
+/obj/item/reagent_containers/glass/filterbowl/AltClick(mob/user)
```
Comment (swindly) [inline code/modules/reagents/reagent_containers/glass.dm]: Add canUseTopic() check?

**Comment by fludd12** [comment]
> Are there any chems (that someone might reasonably ghetto craft) where it will be difficult to actually key a filter to it because the chem ends being a small amount compared to the rest of the reagents in the container unless you use very precise amounts?

I don't _think_ so, but even if there are, you could simply use multiple filters. Since filters are just paper, it is very easy to acquire more. Notably, the filters themselves do not identify the reagent they filter - they simply turn the same color as the reagent. As a result, you may need to do some testing to figure out which reagent is which.

File: code/modules/reagents/reagent_containers/glass.dm (hunk 1 above)
Comment (fludd12) [reply code/modules/reagents/reagent_containers/glass.dm]: Good call!

**Comment by ninjanomnom** [comment]
Having to deal with dirty reagent mixes sounds like exactly the sort of thing that should be a downside of makeshift setups. I'm not a fan of this at all, especially with how this can work with any reagent which makes no sense.

**Comment by fludd12** [comment]
It's not supposed to make any realistic sense, but then again, not a lot in this game does - especially our chemistry system. At the very least, using paper filters is somewhat intuitive, given the coffee example above.

Regarding the balance change to makeshift chemistry, this does have some drawbacks, namely the need to acquire even more resources if you want pure mixtures, as well as the extra time taken to filter things and figure out which filter is keyed to which chemical. I understand that you may not consider these valid drawbacks compared to the existing ones, but please consider giving it a chance.

**Comment by ninjanomnom** [comment]
After discussing it a bit the issue is more to do with how broad this filtering is. If you want to make different filters for different kinds of reagents with some of those being possible to make with ghetto resources then that would be fine but being able to filter anything with some paper you found is a step too far.

**Comment by Mickyan** [comment]
No idea how difficult this could be to code but one idea for balancing would be to make the paper filter have a margin of error of say, 5 units

Ex. you have a mixture of 30 water and 20 oil, filtering will result in the first container having 25 water and 5 oil, the other 5 water and 15 oil

**Comment by ninjanomnom** [comment]
Feel free to make a new pr with a similar concept this implementation is just too much

**Comment by fludd12** [comment]
Valid, thank you for considering it despite everything.

**Comment by ExcessiveUseOfCobblestone** [comment]
I wouldn't mind seeing filtering, but I would prefer it be based on some sort of property within the reagent (overlap between reagents) vs. magic paper/bowl.

chem should be overall more dirtier, so having a way to cleanly get reagents in terms of as-is implementation is something i'm not really interested in.

**Pull request closed without merging**